python/ray/external_storage.py

```diff
@@ -90,6 +103,119 @@ def restore_spilled_objects(self, keys):
                 # read remaining data to our buffer
                 self._put_object_to_store(metadata, buf_len, f, ref)
 
+    def validate(self):
+        if not os.path.exists(self.directory_path):
+            raise ValueError("The give directory path to store objects, "
+                             f"{self.directory_path}, doesn't exist.")
+
+
+class S3Storage(ExternalStorage):
+    """The class for AWS S3.
+
+    Args:
+        bucket_name(str): Name of the bucket to store objects.
+        prefix(str): Prefix of objects that are stored.
+        get_config_override(dict): Configuration dict that will override
+            the default boto3.resource('s3').Object(bucket_name, key).get()
+        put_config_override(dict): Configuration dict that will override
+            the default boto3.resource('s3').Object(bucket_name, key).put()
+    Raises:
+        RayError if it fails to setup a S3 client. For example, if boto3 is
+        not downloaded, it raises an RayError. It can also raise S3 related
+        exceptions.
+    """
+
+    def __init__(self,
+                 bucket_name: str,
+                 prefix: str = "ray_spilled_object_",
+                 get_config_override: dict = None,
+                 put_config_override: dict = None):
+        try:
+            import boto3
```

Comment:
Have you considered using https://pypi.org/project/smart-open/ or https://pypi.org/project/s3fs/ instead of boto directly? They might implement a bunch of optimizations already.

Reply:
I haven't. I will take a look at them.

Reply:
Thanks! Maybe we can try a simple test of upload performance of a large object (e.g., 100MB) to compare.

Reply:
Sure! Btw, the biggest bottleneck now is not the s3 performance itself, but the fact that it is not parallelized (because each restore object is running on a main thread, and the whole thread is blocked until s3 ops are done. So the second IPC from object manager is blocked while we are waiting for the first one to be done). That should still be handled by a different pr, but I will see if I can utilize these libraries.

Reply:
To parallelize, we should parallelize cpp -> cython -> python -> cython -> cpp -> reply path, which was not easy to be achieved by just using threadpool in python code

Reply:
Right-- I was thinking parallelism can be handled by having multiple IO workers. Also, if we coalesce many objects into big blocks, then the only thing that matters is large-chunk write performance.

Reply:
Let me do some profiling with large enough object size (I assume 100MB?) first. I still feel like io worker level parallelization is not good enough (given how slow it is now). So, I will

1. Try smart open with sample test (benchmark) 
2. Profile s3 put/get
3. Write a One pager design doc to compare io worker vs io worker/thread pool vs io worker/event loop implementation.

Reply:
Maybe 1 by this sprint and 2, 3 in the next

Reply:
Ok. I read some doc/source code of both projects, and smart open seems to be more mature and has more options (one great part is it supports all major cloud providers). One drawback is that it doesn't really have asyncio compatible APIs (but if we only rely on Process-level parallelization or thread pool, it should be fine).

Reply:
boto3 actually supports parallelism by multipart uploading.

Reply:
@suquark How big size objects are supposed to be to make it effective? Btw, parallelization I mentioned here is to context switching to the next download code while we are waiting for network IO.